gms/gossiper.cc

```diff
@@ -731,12 +740,69 @@ future<> gossiper::do_status_check() {
     }
 }
 
-future<gossiper::endpoint_permit> gossiper::lock_endpoint(inet_address ep) {
-    return _endpoint_locks.get_or_load(ep, [] (const inet_address& ep) { return semaphore(1); }).then([] (auto eptr) {
-        return get_units(*eptr, 1).then([eptr] (auto units) mutable {
-            return endpoint_permit{std::move(eptr), std::move(units)};
-        });
-    });
+gossiper::endpoint_permit::endpoint_permit(endpoint_locks_map::entry_ptr&& ptr, inet_address addr, std::string caller) noexcept
+    : _ptr(std::move(ptr))
+    , _permit_id(_ptr->pid)
+    , _addr(std::move(addr))
+    , _caller(std::move(caller))
+{
+    ++_ptr->holders;
+    logger.debug("{}: lock_endpoint {}: acquired: permit_id={} holders={}", _caller, _addr, _permit_id, _ptr->holders);
+}
+
+gossiper::endpoint_permit::endpoint_permit(endpoint_permit&& o) noexcept
+    : _ptr(std::exchange(o._ptr, nullptr))
+    , _permit_id(std::exchange(o._permit_id, null_permit_id))
+    , _addr(std::exchange(o._addr, inet_address{}))
+    , _caller(std::move(o._caller))
+{}
+
+gossiper::endpoint_permit::~endpoint_permit() {
+    release();
+}
+
+bool gossiper::endpoint_permit::release() noexcept {
+    if (auto ptr = std::exchange(_ptr, nullptr)) {
+        assert(ptr->pid == _permit_id);
+        logger.debug("{}: lock_endpoint {}: released: permit_id={} holders={}", _caller, _addr, _permit_id, ptr->holders);
+        if (!--ptr->holders) {
+            logger.debug("{}: lock_endpoint {}: released: permit_id={}", _caller, _addr, _permit_id);
+            ptr->units.return_all();
+            ptr->pid = null_permit_id;
+            _permit_id = null_permit_id;
+            return true;
+        }
+    }
+    return false;
+}
+
+gossiper::endpoint_lock_entry::endpoint_lock_entry() noexcept
+    : sem(1)
+    , pid(permit_id::create_null_id())
+{}
+
+future<gossiper::endpoint_permit> gossiper::lock_endpoint(inet_address ep, permit_id pid, seastar::compat::source_location l) {
```

Comment:
Please document when the caller of lock_endpoint needs to pass permit_id and when it is not needed.

Reply:
okay (in a follow-up)